src/sentry/grouping/strategies/newstyle.py

```diff
@@ -342,9 +342,63 @@ def get_contextline_component(frame, platform, function, context):
     return component
 
 
-@strategy(id="stacktrace:v1", interfaces=["stacktrace"], variants=["!system", "app"], score=1800)
+@strategy(id="stacktrace:v1", interfaces=["stacktrace"], score=1800)
 def stacktrace(stacktrace, context, **meta):
+    assert context["variant"] is None
+
+    if context["hierarchical_grouping"]:
+        rv = call_with_variants(
+            _single_stacktrace_variant,
+            # when app hash is equal to system hash, we do not want to stop it
+            # from contributing as it will become a hierarchical hash when
+            # renamed to app-depth-max. Therefore we must not make system a
+            # mandatory variant ('system' instead of '!system')
+            ["system", "app"],
+            stacktrace,
+            context=context,
+            meta=meta,
+        )
+
+        full_stacktrace = rv.pop("app")
+        rv["app-depth-max"] = full_stacktrace
+
+        for max_frames in range(1, 6):
+            stacktrace = full_stacktrace.shallow_copy()
+            new_values = []
+            ignored_frames = 0
+
+            # cannot update contributes here as this copy is shallow.
+            # instead, trim down list
+            for component in reversed(stacktrace.values):
+                if not component.contributes:
+                    continue
+
+                if len(new_values) < max_frames:
+                    new_values.append(component)
+                else:
+                    ignored_frames += 1
+
+            if not new_values or not ignored_frames:
+                break
+
+            new_values.reverse()
+            stacktrace.update(values=new_values)
+
+            rv[f"app-depth-{max_frames}"] = stacktrace
+
+            max_frames += 1
```

Comment:
This line has no effect (I think). Was this refactored from a while loop?

Reply:
thanks, yes